src/debug-generator/debug-generator.c

```diff
@@ -145,6 +187,59 @@ static int install_debug_shell_dropin(void) {
         const char *tty = arg_debug_tty ?: arg_default_debug_tty;
         int r;
 
+        if (arg_boot_breakpoint != _BOOT_BREAKPOINT_TYPE_INVALID) {
+                switch (arg_boot_breakpoint) {
+                case BOOT_BREAKPOINT_PRE_UDEV:
+                        r = write_drop_in_format(arg_dest, "debug-shell.service", 60, "breakpoint",
+                                                 "# Automatically generated by systemd-debug-generator\n\n"
+                                                 "[Unit]\n"
+                                                 "Wants=systemd-journald.socket\n"
+                                                 "After=systemd-journald.socket\n"
+                                                 "Before=systemd-udevd.service systemd-udev-trigger.service\n"
+                                                 "\n[Service]\n"
+                                                 "Environment=PS1=\"%s:$PWD# \"\n"
+                                                 "Type=oneshot\n"
+                                                 "Restart=no\n",
+                                                 boot_breakpoint_type_to_string(arg_boot_breakpoint));
+                        break;
+
+                case BOOT_BREAKPOINT_PRE_SYSROOT_MOUNT:
+                        assert(in_initrd());
+                        r = write_drop_in_format(arg_dest, "debug-shell.service", 60, "breakpoint",
+                                                 "# Automatically generated by systemd-debug-generator\n\n"
+                                                 "[Unit]\n"
+                                                 "After=sysinit.target\n"
+                                                 "Before=initrd-root-fs.target sysroot.mount systemd-fsck-root.service\n"
+                                                 "\n[Service]\n"
+                                                 "Environment=PS1=\"%s:$PWD# \"\n"
+                                                 "Type=oneshot\n"
+                                                 "Restart=no\n",
+                                                 boot_breakpoint_type_to_string(arg_boot_breakpoint));
+                        break;
+
+                case BOOT_BREAKPOINT_PRE_SWITCH_ROOT:
+                        assert(in_initrd());
+                        r = write_drop_in_format(arg_dest, "debug-shell.service", 60, "breakpoint",
```

Comment:
Frankly, I don't think reusing debug-shell.service here is a good idea. It spawns shell at a different tty, and the service is designed to be stick around forever, but now you're inserting deps via drop-in, which affects the use case where both generic debug shell and breakpoint is utilized.

Just use something like emergency.service?

Reply:
I think it is better to introduce a completely new service for the usecase, rather than reusing existing service.
Then, we should allow to configure the multiple break point services.

So, rather than dynamically creating drop-in, I think it is better to create static service unit files for each break points, and simply enable them by the kernel command line switch.

Reply:
> which affects the use case where both generic debug shell and breakpoint is utilized.

I'm not sure if that would be a use case for many people, but ok.

> Just use something like emergency.service?

Hmm, I'm not sure about that...

> I think it is better to introduce a completely new service for the usecase, rather than reusing existing service.

That would be ok.

> So, rather than dynamically creating drop-in, I think it is better to create static service unit files for each break points, and simply enable them by the kernel command line switch.

That would be ok, but... do we need a generator for that? wouldn't it be enough to simply have services with `ConditionKernelCommandLine=...`?

Reply:
> That would be ok, but... do we need a generator for that? wouldn't it be enough to simply have services with `ConditionKernelCommandLine=...`?

Indeed. If that works, then OK for me. @poettering WDYT?

Reply:
> > So, rather than dynamically creating drop-in, I think it is better to create static service unit files for each break points, and simply enable them by the kernel command line switch.
> 
> That would be ok, but... do we need a generator for that? wouldn't it be enough to simply have services with `ConditionKernelCommandLine=...`?

See https://github.com/systemd/systemd/blob/156a77ddbc09a066a387e8ee41124d2b00fecd43/src/rc-local-generator/rc-local-generator.c#L15C1-L19C32:
```
/* So you are reading this, and might wonder: why is this implemented as a generator rather than as a plain, statically
 * enabled service that carries appropriate ConditionFileIsExecutable= lines? The answer is this: conditions bypass
 * execution of a service's binary, but they have no influence on unit dependencies. Thus, a service that is
 * conditioned out will still act as synchronization point in the dependency tree, and we'd rather not have that for
 * these two legacy scripts. */
```
IMHO the same reasoning applies here. The proposed breakpoint units are for occasional use, therefore they shouldn't affect normal boot in any way.

Reply:
Ahh thanks for this background info. Then I'll rework this as suggested in https://github.com/systemd/systemd/pull/35410#discussion_r1863478648, adding also a similar comment for future readers.